src/chainbin/chain-make-num-fst-e2e.cc

```diff
@@ -0,0 +1,121 @@
+// chainbin/chain-make-num-fst-e2e.cc
+
+// Copyright 2019  Johns Hopkins University (author: Yiwen Shao)
+
+// See ../../COPYING for clarification regarding multiple authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
+// WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+// See the Apache 2 License for the specific language governing permissions and
+// limitations under the License.
+
+/** @brief Converts fsts (containing transition-ids) to fsts (containing pdf-ids, zero-based).
+*/
+#include "base/kaldi-common.h"
+#include "gmm/am-diag-gmm.h"
+#include "hmm/transition-model.h"
+#include "hmm/hmm-utils.h"
+#include "util/common-utils.h"
+#include "fst/fstlib.h"
+
+namespace kaldi {
+
+bool FstTransitionToPdf(const fst::StdVectorFst &fst_transition,
+			const TransitionModel &trans_model,
+			fst::StdVectorFst *fst_pdf) {
+  fst::StdVectorFst fst_tmp(fst_transition);
+  fst::RemoveEpsLocal(&fst_tmp);
+  fst::RmEpsilon(&fst_tmp);
+  // first change labels to pdf-id + 1
+  int32 num_states = fst_tmp.NumStates();
+  for (int32 state = 0; state < num_states; state++) {
+    for (fst::MutableArcIterator<fst::StdVectorFst> aiter(&fst_tmp, state);
+         !aiter.Done(); aiter.Next()) {
+      const fst::StdArc &arc = aiter.Value();
+      if (arc.ilabel == 0) {
+        KALDI_WARN << "Utterance rejected due to eps on input label";
+        return false;
+      }
+      KALDI_ASSERT(arc.ilabel != 0);
+      fst::StdArc arc2(arc);
+      arc2.ilabel = arc2.olabel = trans_model.TransitionIdToPdf(arc.ilabel) + 1;
+      aiter.SetValue(arc2);
+    }
+  }
+  *fst_pdf = fst_tmp;
+  return true;
+}
+
+bool AddWeightToFst(const fst::StdVectorFst &normalization_fst,
+		    fst::StdVectorFst *fst) {
+  // Note: by default, 'Compose' will call 'Connect', so if the
+  // resulting FST is not connected, it will end up empty.
+  fst::StdVectorFst composed_fst;
+  fst::Compose(*fst, normalization_fst,
+	       &composed_fst);
+  *fst = composed_fst;
+  if (composed_fst.NumStates() == 0)
+    return false;
+  return true;
+}
+
+}
+
+int main(int argc, char *argv[]) {
+  using namespace kaldi;
+  typedef kaldi::int32 int32;
+  try {
+    const char *usage =
+      "Converts chain e2e numerator fst (containing transition-ids) to fst (containing pdf-ids, \n"
+      "zero-based, eps-removed, and normalized by the normalization fst) \n"
```

Comment:
here is where it says zero-based.  Also this doesn't remove epsilons, it just checks that there are no epsilons.. might be good to clarify.   should say pdf-ids + 1.